internal/ovsexporter/datapath.go

```diff
@@ -0,0 +1,177 @@
+// Copyright 2018 DigitalOcean.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ovsexporter
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/digitalocean/go-openvswitch/ovsnl"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var _ prometheus.Collector = &datapathCollector{}
+
+// A datapathCollector is a prometheus.Collector for Open vSwitch datapaths.
+type datapathCollector struct {
+	Info                       *prometheus.Desc
+	StatsHitsTotal             *prometheus.Desc
+	StatsMissesTotal           *prometheus.Desc
+	StatsLostTotal             *prometheus.Desc
+	StatsFlows                 *prometheus.Desc
+	MegaflowStatsMaskHitsTotal *prometheus.Desc
+	MegaflowStatsMasksTotal    *prometheus.Desc
+
+	listDatapaths func() ([]ovsnl.Datapath, error)
+}
+
+// newDatapathCollector creates a prometheus.Collector for Open vSwitch
+// datapaths.  The input function can be swapped for testing.
+func newDatapathCollector(fn func() ([]ovsnl.Datapath, error)) prometheus.Collector {
+	const (
+		subsystem = "datapath"
+	)
+
+	var (
+		labels = []string{"datapath"}
+	)
+
+	return &datapathCollector{
+		Info: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, subsystem, "info"),
+			"Metadata about a given datapath.",
+			// TODO(mdlayher): expose features?
```

Comment:
what features would these be?

Reply:
Yeah so it looks like some netlink implementation detail stuff that I'm honestly not sure about, so my gut feeling is that we should omit them until we find some use for them.

Reply:
I'll probably just remove this info metric and Todo for now.